### DISCUSSION TITLE
Generators fields and custom attributes

### DISCUSSION THREAD

**Discussion by maoosi** [prisma/prisma#2514] [Q&A] [answered]
**Hey everyone!**

I’ve recently started working on a Prisma 2.0 Generator that automatically generates "deploy-ready" code to create a GraphQL [OpenCRUD](https://www.opencrud.org/) API on AWS AppSync. The idea is to facilitate the development and maintenance of an API using Prisma and the AWS serverless platform.
 
```graphql
# schema.prisma
generator appsync {
    provider = "prisma-appsync" # my generator
    output = "./src" # output dir for the generator
    # authorization = "./authorization-modes.js"
}
```
 
It's still a big work in progress, but the generated output will look like this once finalised (all the files generated are designed to be shipped via AWS CDK):

- **Schema**: GraphQL Schema file including [AppSync defined scalars](https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html).
- **Resolvers**: JavaScript file that exports all the resolver mapping templates [AWS AppSync Lambda resolver format](https://docs.aws.amazon.com/appsync/latest/devguide/resolver-mapping-template-reference-lambda.html).
- **Data source**: TypeScript class that exposes all relevant functions for the CRUD API (Using Prisma Client and fully extensible).
- **Boilerplate**: Package including a sample Lambda function (default setup implementing the CRUD API class) and a sample AWS CDK file to deploy the AppSync API.
 

One thing that I'm trying to work out, is the best way to implement [AppSync authorization modes](https://docs.aws.amazon.com/appsync/latest/devguide/security.html#using-additional-authorization-modes) against specific types and/or fields (related to the Prisma models and fields). I was able to make it work by adding a custom field into my generator block `authorization = "./authorization-modes.js"`. The problem with that is that it's adding complexity, as well as a secondary source of truth for the Schema rather than keeping everything contained within the Prisma Schema file.
 
```javascript
// authorization-modes.js
export default {
    "models": [
        {
            "name": "Post",
            "directives": ["@awsapikey", "@awsiam"],
            "fields": [
                {
                    "name": "title",
                    "directives": ["@awsiam"]
                }
            ]
        }
    ]
}
```
 
**Questions related to the above:**

- **[1]** Is the ability to add extra fields into generator blocks intentional? Or is it something that might be removed in future releases since not mentioned in the docs?

- **[2]** Is there any plan for allowing people to add custom attributes for both models and fields? For my use case, something like `@@params(_ [String]: [String])` would allow using `@@params(_ directives: "@awsiam")` and reading it from the Generator Handler. I'm guessing it's something that could become handy for people who would like to extend Prisma similarly by creating custom Generators.
 

Thanks,
Sylvain

**Comment by nikolasburk** [comment]
Hey @maoosi, first of all huge kudos to you, this looks like an awesome project, I'm excited to see how it evolves 😄 

Regarding your questions, to be honest we don't fully know yet but your use case is absolutely valid and we do want to somehow enable folks to "extend" the Prisma schema when custom generators are used (could e.g. also be via some sort of plugin system). The best thing would be if you could open up feature requests to persist this scenario and make sure we can properly incorporate it into our planning!

By the way, did you already see the [Prisma schema spec](https://github.com/prisma/specs/tree/master/schema)? This might also include a few pointers regarding your questions (note though that the spec is also still subject to change, so things might turn out different than specified at the moment).

**Comment by maoosi** [comment, marked as answer]
Thanks @nikolasburk, I’ll make the project public on Github after I have all the pieces working nicely together.

I just had a look at the Prisma schema spec that you've shared, and I could find an answer to my first question. It seems like that adding extra fields into generator blocks is indeed intentional. That’s good to know, as it will help with my use case.

I’ll definitely open a feature request around the idea of extending the Prisma Schema / adding attributes for custom generators. Keen to see that coming along.

**Reply by nikolasburk** [reply]
Hey @maoosi  👋 I just wanted to check in with you! Are you still working on this? Anything we can do to help? 🙂

**Reply by maoosi** [reply]
Hey @nikolasburk, thanks for asking. Yes, still working on this!

I ended up using comments in the Prisma.schema file to manage appsync directives. As for granular authorization rules, they can be applied directly via the API Class that's being created by the generator.

It’s already powering [travistravis.co](https://travistravis.co) and another project that I’m working on, but will still require a bit of effort to make it work across different use cases, write some tests, document it and simplify overall usage as much as possible. I’ll post an update here once I’m ready to publish something.

**Reply by SrZorro** [reply]
Hello, looks like you took the same path we are taking, using the prisma comments to implement on top of it a way to add attributes. We have an [open RFC in this issue](https://github.com/MichalLytek/typegraphql-prisma/issues/35) with a proposal to resolve the validation and authorization for that library building a plugin system for Prisma. That package implements an autogenerated CRUD system with [TypeGraphQL](https://typegraphql.com/) + Prisma, and we are building a system to extend the CRUD with plugins that let you add attributes to `prisma.schema` like validation and authorization.

Maybe could you @maoosi check it out and share some of the knowlage you probably have on the subject?

Also @nikolasburk we would hope that maybe this implementation we are doing there could help as a steping stone for a posible plugin system for Prisma, so we would like to get feedback from the Prisma team so the work we are doing there could be translated to a native implementation for Prisma and use this project as a playground for a posible implementation for plugin consumers.

**Reply by matthewmueller** [reply]
Hey @SrZorro custom attributes is definitely something on our mind. We're trying to wrap up a couple other topics at the moment (group by, transactions), but then we want to look into Prisma Schema extensibility and validation :)

**Reply by maoosi** [reply]
Hey @SrZorro, I’ve had a look at the open RFC and it is very close to what I’ve put in place inside my own generator.

Here's an overview of a `prisma.schema` file using AST comments to help with authorization:

```prisma
/// @PrismaAppSync.type: '@aws_cognito_user_pools(cognito_groups: ["admins"])'
model Post {
    id          Int         @id @default(autoincrement())
    title       String
    updatedAt   DateTime    @updatedAt
}
```

And here is the output from my generator:

```graphql
type Post @aws_cognito_user_pools(cognito_groups: ["admins"]) {
    id: Int!
    title: String
    updatedAt: AWSDateTime!
}
```

I haven't released anything yet (still refactoring certain things and working on docs), but you can have a look at [the work in progress docs](https://prisma-appsync.vercel.app/usage/authorization.html) that goes into more details around Authorization.

To parse the AST comments from the Prisma DMMF object, I've used the following package which make it easy extracting JavaScript objects from dot annotation strings: [Akryum/graphql-annotations](https://github.com/Akryum/graphql-annotations).

Feel free to ask if you have any particular question on this.

**Comment by ludralph** [comment]
Hi @maoosi,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
